src/backend/access/external/libchurl.c

```diff
@@ -312,16 +313,29 @@ CHURL_HANDLE churl_init(const char* url, CHURL_HEADERS headers)
 
 	/* needed to resolve localhost */
 	if (strstr(url, LocalhostIpV4) != NULL) {
-		struct curl_slist *resolve_hosts = NULL;
-		char *pxf_host_entry = (char *) palloc0(strlen(pxf_service_address) + strlen(LocalhostIpV4Entry) + 1);
-		strcat(pxf_host_entry, pxf_service_address);
-		strcat(pxf_host_entry, LocalhostIpV4Entry);
-		resolve_hosts = curl_slist_append(NULL, pxf_host_entry);
-		set_curl_option(context, CURLOPT_RESOLVE, resolve_hosts);
-		pfree(pxf_host_entry);
-	}
+		//get loopback interface ip address
+		char* loopback_addr = get_loopback_ip_addr();
+		elog(DEBUG1, "Loopback interface IP address: %s", loopback_addr);
+		//find host start position in url
+		char* start = strstr(url, LocalhostIpV4);
+		//allocate new url with replaced host
+		char* replaced_url = palloc(strlen(url) + strlen(loopback_addr) - strlen(LocalhostIpV4) + 1);
+		//token before host
+		char* before_host = pnstrdup(url, start - url);
+		//token after host
+		char* after_host = pstrdup(url + (start - url) + strlen(LocalhostIpV4));
+		//construct replaced url using loopback interface ip
+		sprintf(replaced_url, "%s%s%s", before_host, loopback_addr, after_host);
```

Comment:
Checked through HAWQ code - haven't found much null checks after pallocating memory. I guess palloc might check for NULL inside.